nca/CoreDS/TcpLikeProperties.py

```diff
@@ -36,42 +38,49 @@ class TcpLikeProperties(CanonicalHyperCubeSet):
         :param MethodSet methods: the set of http request methods
         :param MinDFA paths: The dfa of http request paths
         :param MinDFA hosts: The dfa of http request hosts
-        :param CanonicalIntervalSet peers: the set of (target) peers
+        :param CanonicalIntervalSet icmp_type: ICMP-specific parameter (type dimension)
+        :param CanonicalIntervalSet icmp_code: ICMP-specific parameter (code dimension)
         :param PeerSet base_peer_set: the base peer set which is referenced by the indices in 'peers'
+        :param CanonicalIntervalSet src_peers: the set of source peers
+        :param CanonicalIntervalSet dst_peers: the set of target peers
         """
         super().__init__(TcpLikeProperties.dimensions_list)
-        assert not peers or base_peer_set
+        assert (not src_peers and not dst_peers) or base_peer_set
 
         self.named_ports = {}  # a mapping from dst named port (String) to src ports interval set
         self.excluded_named_ports = {}  # a mapping from dst named port (String) to src ports interval set
         self.base_peer_set = base_peer_set if base_peer_set else PeerSet()
+        if create_empty:
+            return
 
         # create the cube from input arguments
         # create a dict object that holds the values required to build the cube
-        dims_to_values = {"src_ports": {"value": source_ports.port_set,
-                                        "is_all": source_ports.is_all()},
-                          "dst_ports": {"value": dest_ports.port_set,
-                                        "is_all": dest_ports.is_all()},
-                          "methods": {"value": methods,
-                                      "is_all": methods.is_whole_range()},
-                          "hosts": {"value": hosts,
-                                    "is_all": hosts is None},
-                          "paths": {"value": paths,
-                                    "is_all": paths is None},
-                          "peers": {"value": peers,
-                                    "is_all": peers is None}}
+        dims_to_values = {"src_peers": {"value": src_peers, "is_all": src_peers is None},
+                          "dst_peers": {"value": dst_peers, "is_all": dst_peers is None},
+                          "protocols": {"value": protocols, "is_all": protocols is None or protocols.is_whole_range()},
+                          "src_ports": {"value": source_ports.port_set, "is_all": source_ports.is_all()},
+                          "dst_ports": {"value": dest_ports.port_set, "is_all": dest_ports.is_all()},
+                          "methods": {"value": methods, "is_all": methods is None or methods.is_whole_range()},
+                          "hosts": {"value": hosts, "is_all": hosts is None},
+                          "paths": {"value": paths, "is_all": paths is None},
+                          "icmp_type": {"value": icmp_type, "is_all": icmp_type is None},
```

Comment:
is it possible that `icmp_type` is an object with the range of the entire domain rather than `None`?

Reply:
Fixed, Also for icmp_code.